### PR TITLE
[BugFix] Write short index page using sort key if table separate primary keys and sort keys

### LINKED ISSUE #21375: [sort key]query result is empty
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
```
CREATE TABLE `test_pk_sort_key4` (
`k1` date NOT NULL,
`k2` datetime NOT NULL,
`k3` varchar(20) NOT NULL,
`k4` varchar(20) NOT NULL,
`k5` boolean NOT NULL,
`v1` tinyint NULL,
`v2` smallint NULL,
`v3` int NULL,
`v4` bigint NULL,
`v5` largeint NULL,
`v6` float NULL,
`v7` double NULL,
`v8` decimal(27,9) NULL
) 
PRIMARY KEY(`k1`, `k2`, `k3`, `k4`, `k5`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 3
ORDER BY(`v1`, `v5`, `v4`, `v3`, `v2`)
PROPERTIES (
"replication_num" = "3",
"storage_format" = "v2"
);

insert into test_pk_sort_key4 values('2020-06-23', '2020-06-23 00:00:27', 'beijing', 'wergergqer', 0, -101, -32741, 1111, 11111, 111111, 11.11, 111.111, 1111.1111);

select v5 from test_pk_sort_key4;
select v5 from test_pk_sort_key4 where v5 = 111111;

select * from test_pk_sort_key4;
select v1,v5 from test_pk_sort_key4 where v5 = 111111;
```
image

### Expected behavior (Required)
result is not empty
### Real behavior (Required)
result is empty
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
mysql> select current_version();
+----------------------------+
| current_version()          |
+----------------------------+
| BRANCH-3.0-RELEASE 9cc2e97 |
+----------------------------+
1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by sevev**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes https://github.com/StarRocks/starrocks/issues/21375 https://github.com/StarRocks/starrocks/issues/21372

## Problem Summary(Required) ：

If primary key table separate primary key and sort key, we will write data in the order of  sort keys. However, we still use primary keys to generate the short key index,  so we may get error scan range and get error result when we execute query.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/storage/seek_tuple.h [hunk 1]
```diff
@@ -80,4 +82,23 @@ inline std::string SeekTuple::short_key_encode(size_t num_short_keys, uint8_t pa
     return output;
 }
 
+inline std::string SeekTuple::short_key_encode(size_t num_short_keys, std::vector<uint32_t> short_key_idxes,
+                                               uint8_t padding) const {
+    std::string output;
+    size_t n = std::min(num_short_keys, std::min(short_key_idxes.size(), _values.size()));
```
Comment (chaoyli) [inline be/src/storage/seek_tuple.h:88]: Under which circumstance, the `num_short_keys` is less than `short_key_idxes`?

File: be/src/storage/seek_tuple.h [hunk 2]
```diff
@@ -80,4 +82,23 @@ inline std::string SeekTuple::short_key_encode(size_t num_short_keys, uint8_t pa
...
+    size_t val_num = _values.size();
+    for (auto i = 0; i < n; i++) {
+        if (short_key_idxes[i] >= val_num || _values[short_key_idxes[i]].is_null()) {
+            output.push_back(KEY_NULL_FIRST_MARKER);
+        } else {
+            output.push_back(KEY_NORMAL_MARKER);
+            _schema.field(short_key_idxes[i])->encode_ascending(_values[short_key_idxes[i]], &output);
+        }
+    }
+    if (_values.size() < num_short_keys) {
```
Comment (chaoyli) [inline be/src/storage/seek_tuple.h:98]: When the `_values.size()` will be less than `num_short_keys`?

File: be/src/storage/seek_tuple.h (hunk 1 above)
Comment (sevev) [reply be/src/storage/seek_tuple.h:88]: > Under which circumstance, the `num_short_keys` is less than `short_key_idxes`?

If some sort columns type is float or double, we will truncate the sort keys and  the `num_short_keys` may be less than the num of`short_key_idxes`?

File: be/src/storage/seek_tuple.h (hunk 2 above)
Comment (sevev) [reply be/src/storage/seek_tuple.h:98]: > When the `_values.size()` will be less than `num_short_keys`?

This scenario may happened if we do partial update for the table which separate primary keys and sort keys. But we don't support partial update for this table right now. So it should never happened right now.

File: be/src/storage/seek_tuple.h (hunk 1 above)
Comment (chaoyli) [reply be/src/storage/seek_tuple.h:88]: Do wen need to prohibit the sort key have `float` or `double`?

File: be/src/storage/seek_tuple.h (hunk 2 above)
Comment (chaoyli) [reply be/src/storage/seek_tuple.h:98]: So the partial update is limited due to lack of this partial update mode.

Comment (sevev) [reply be/src/storage/seek_tuple.h:98]: > So the partial update is limited due to lack of this partial update mode.

Yes, table with sort key do not support partial update now.

**Comment by sevev** [comment]
@chaoyli @decster All comments are resolved, PTAL~

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 29 (72.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|12|58.33%|[742, 743, 744, 745, 746]|
|:large_blue_circle:|com/starrocks/analysis/KeysDesc.java|5|7|71.43%|[247, 248]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|9|10|90.00%|[164]|

File: be/src/storage/seek_tuple.h (hunk 1 above)
Comment (chaoyli) [inline be/src/storage/seek_tuple.h:88]:
```suggestion
    DCHECK(num_short_keys == short_key_idxes.size());
    size_t n = std::min(num_short_keys, _values.size());
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 29 (72.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|12|58.33%|[742, 743, 744, 745, 746]|
|:large_blue_circle:|com/starrocks/analysis/KeysDesc.java|5|7|71.43%|[247, 248]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|9|10|90.00%|[164]|

**Review by chaoyli — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 29 (72.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|12|58.33%|[742, 743, 744, 745, 746]|
|:large_blue_circle:|com/starrocks/analysis/KeysDesc.java|5|7|71.43%|[247, 248]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|9|10|90.00%|[164]|

**Review by decster — DISMISSED** [review]

File: be/src/storage/seek_tuple.h [hunk 3]
```diff
@@ -80,4 +82,23 @@ inline std::string SeekTuple::short_key_encode(size_t num_short_keys, uint8_t pa
     return output;
 }
 
+inline std::string SeekTuple::short_key_encode(size_t num_short_keys, std::vector<uint32_t> short_key_idxes,
```
Comment (chaoyli) [inline be/src/storage/seek_tuple.h:85]: sort_key_indxes

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 29 (72.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|12|58.33%|[742, 743, 744, 745, 746]|
|:large_blue_circle:|com/starrocks/analysis/KeysDesc.java|5|7|71.43%|[247, 248]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|9|10|90.00%|[164]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 29 (72.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|12|58.33%|[742, 743, 744, 745, 746]|
|:large_blue_circle:|com/starrocks/analysis/KeysDesc.java|5|7|71.43%|[247, 248]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|9|10|90.00%|[164]|

File: be/src/storage/seek_tuple.h (hunk 3 above)
Comment (sevev) [reply be/src/storage/seek_tuple.h]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 29 (72.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|12|58.33%|[742, 743, 744, 745, 746]|
|:large_blue_circle:|com/starrocks/analysis/KeysDesc.java|5|7|71.43%|[247, 248]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|9|10|90.00%|[164]|

File: be/src/storage/seek_tuple.h [hunk 4]
```diff
@@ -81,4 +83,24 @@ inline std::string SeekTuple::short_key_encode(size_t num_short_keys, uint8_t pa
     return output;
 }
 
+inline std::string SeekTuple::short_key_encode(size_t num_short_keys, std::vector<uint32_t> sort_key_idxes,
+                                               uint8_t padding) const {
+    std::string output;
+    DCHECK(num_short_keys <= sort_key_idxes.size());
+    size_t n = std::min(num_short_keys, std::min(sort_key_idxes.size(), _values.size()));
```
Comment (chaoyli) [inline be/src/storage/seek_tuple.h:90]:
```suggestion
    size_t n = std::min(num_short_keys, _values.size());
```

**Review by chaoyli — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**